compat/openssl.h

```diff
@@ -0,0 +1,241 @@
+/*
+ * Copyright (C) 1996-2018 The Squid Software Foundation and contributors
+ *
+ * Squid software is distributed under GPLv2+ license and includes
+ * contributions from numerous individuals and organizations.
+ * Please see the COPYING and CONTRIBUTORS files for details.
+ */
+
+/* OpenSSL API changed dramatically between 1.0.2 and 1.1.0, and
+ * compatibility was broken.  Most of the structures became opaque,
+ * and access functions were created.  There's no (safe) way to
+ * access the struct members any more, so the solution is to use
+ * the new API in the main code, and add the functions for older
+ * versions in compat/openssl.h.
+ * Once all the supported library versions use the new API, the shim
+ * can be dropped.
+ */
+
+#ifndef OPENSSL_COMPAT_H
+#define OPENSSL_COMPAT_H
+
+#if !USE_OPENSSL
+#error compat/openssl.h depends on USE_OPENSSL
+#endif
+
+#if defined(__cplusplus)
```

Comment:
The two conditions above are true.

The code is not compatible with C as is.  Besides the use of a couple of `nullptr` in the code (OK, this is trivial), there is the difference in semantics for `inline` between C89 (no inline at all), GNU89, C99 and C++ that's a bit awkward to accommodate.  If you leave it as is and compile it under C99 rules, the compiler will generate warnings about missing declarations, but should work with optimization.  It took me a while to understand what was going on.  To make it portable, an `extern inline` definition must exist somewhere, so that the function can be called if it is not actually inlined.  

What I think it's the best solution is to use `defined(__cplusplus)` to assert that `compat/openssl.h` is being used only in C++.  If there is ever a need to use it in C, then the extern inline definitions can be created.  I'll add a note to the assertion as a guide.

Reply:
Got it.